### PR TITLE
expression: rewrite builtin function: CHAR_LENGTH

### PR DESCRIPTION
**Pull request by spongedu**
for #4080

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -1502,33 +1502,26 @@ type charLengthFunctionClass struct {
 }
 
 func (c *charLengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinCharLengthSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
```
Comment (zz-jason) [inline expression/builtin_string.go]: `verifyArgs(args)` firstly

File: expression/builtin_string.go [hunk 2]
```diff
@@ -1502,33 +1502,26 @@ type charLengthFunctionClass struct {
...
+	sig := &builtinCharLengthSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinCharLengthSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
 // eval evals a builtinCharLengthSig.
```
Comment (zz-jason) [inline expression/builtin_string.go]: update comment

**Comment by zz-jason** [comment]
- add test in `expression/integration_test.go`
- add test in `expression/builtin_string_test.go:TestCharLength` to test `f.isDeterministic()`

File: expression/builtin_string.go (hunk 2 above)
Comment (spongedu) [inline expression/builtin_string.go]: ok

File: expression/builtin_string.go (hunk 1 above)
Comment (spongedu) [inline expression/builtin_string.go]: done. It seems that 'verifyArgs(args)' is not done first in many builtin functions. 'LENGTH' for example. Maybe need refine

Comment (zz-jason) [inline expression/builtin_string.go]: yes, currently, `verifyArgs` only check argument count, you can pull another request to refine them.

Comment (zz-jason) [inline expression/builtin_string.go]: related to this issue: https://github.com/pingcap/tidb/issues/4106

**Comment by breeswish** [comment]
LGTM

**Pull request merged**